### ISSUE TITLE
Eio.Net.getaddrinfo_stream fails to resolve host after 1019 successful calls on linux

### ISSUE THREAD

**Issue by serpent7776** [ocaml-multicore/eio#776]
I'm writing an application using cohttp-eio that repeatedly connects to a remote host every 3s. The initial 1019 requests succeed, but all the following requests fail to resolve the host. I can simplify the code to the following:

```ocaml
let connect ~sw ~env addr =
  ignore @@ Eio.Net.connect ~sw env#net addr

let print addr =
  Format.printf "@[%a@]@." Eio.Net.Sockaddr.pp addr

let connect_and_print ~sw ~env addr =
  print addr;
  (match addr with | `Tcp (a,_) -> Eio.Net.Ipaddr.fold ~v4:(fun _ -> connect ~sw ~env addr) ~v6:(fun _ -> ()) a | _ -> ())

let getaddrinfo ~sw ~env uri =
  List.iter (connect_and_print ~sw ~env) @@
  Eio.Net.getaddrinfo_stream ~service:"http" env#net
    (Uri.host_with_default uri)

let google_uri = Uri.of_string "http://google.com"

let rec loop ~sw ~env clock count =
  Format.printf "%d:@." count;
  getaddrinfo ~sw ~env google_uri ;
  loop ~sw ~env clock @@ (count + 1)

let () =
  Eio_main.run @@ fun env ->
  Eio.Switch.run @@ fun sw ->
  let clock = env#clock in
  loop ~sw ~env clock 0
```

This initially given the correct output

```
0:
tcp:142.250.186.206:80
tcp:[2a00:1450:401b:80d::200e]:80
1:
tcp:142.250.186.206:80
tcp:[2a00:1450:401b:80d::200e]:80
```

But after 1019 requests `Eio.Net.getaddrinfo_stream` returns no addresses:

```
1018:
tcp:142.250.186.206:80
tcp:[2a00:1450:401b:80d::200e]:80
1019:
1020:
1021:
1022:
1023:
1024:
1025:
```

I get the same behaviour every single time. I tested this on three different machines: Manjaro, Ubuntu and FreeBSD. It only reproduces on linux systems, so it's likely an issue in `eio_linux`.
Also, it reproduces only when `Eio.Net.connect` is called.

Possibly related https://github.com/ocaml-multicore/eio/issues/351

**Comment by patricoferris** [comment]
I think the problem is due to number of _open_ connections. For example:

```ocaml
let connect ~sw ~env addr =
  let c = Eio.Net.connect ~sw env#net addr in
  Eio.Flow.close c
```

fixes this problem. Note with the use of the single switch these connections are never being closed (that may or may not be desired). Although I'm not quite sure why that is the case on Linux. But just thought I would send this comment to hopefully unblock you.

**Comment by serpent7776** [comment]
Thanks, that's very helpful.
I can confirm that closing the connections manually fixes the issue. I also did a roughly similar code in C and it also has the same behaviour when connections are not closed. After 1021 connections, `getaddrinfo` returns `error: Servname not supported for ai_socktype`, so this is linux behaviour.

But now I'm not sure, if this is eio issue, cohttp issue or am I supposed to close the connections manually?

**Comment by talex5** [comment]
If you're using cohttp-eio, it will close the connection for you as soon as the switch goes out of scope. Just make sure you're not sharing a single switch for everything (like your example does) - see https://github.com/ocaml-multicore/eio?tab=readme-ov-file#switches-1.

The behaviour of getaddrinfo ignoring the error and returning an empty list is questionable, but comes from OCaml's `Unix.getaddrinfo`. It might be better to have our own version that passes errors on.

**Comment by serpent7776** [comment]
Oh, I am using a single switch everywhere :)

**Comment by SGrondin** [comment]
@serpent7776 Switches are cheap, so don't hesitate to create more than one.
Ideally, you should have one Switch per "lifecycle".
If your DNS calls never overlap, then create one Switch per call so each call gets "cleaned up" as soon as possible.
If you have 2 different resources (let's say a file and a DNS call) and they both become "cleanable" (or "closeable") simultaneously, then you can use the same Switch for both.
I think it's often a good idea to create one Switch per "cleanup point". Resources that don't share the same cleanup point should not share the same Switch.

**Comment by talex5** [comment]
We now forward any error from `getaddrinfo` by raising an exception instead of ignoring it.

However, on my system, `getaddrinfo` itself fails to forward the error!

Test-case:
```c
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

int main(int argc, char *argv[])
{
    int              s;
    struct addrinfo  *result;
    struct addrinfo  hints = {
	    .ai_flags = AI_CANONNAME,
    };

    while (1) {
	    if (open("/dev/null", O_RDONLY, 0) < 0) {
		    printf("open: %s\n", strerror(errno));
		    break;
	    }
    }

    s = getaddrinfo("www.google.com", NULL, &hints, &result);
    if (s != 0) {
        fprintf(stderr, "getaddrinfo: %s\n", gai_strerror(s));
        exit(EXIT_FAILURE);
    }

    printf("Got result: %s\n", result->ai_canonname);

    exit(EXIT_SUCCESS);
}
```
On Debian 13, I get:
```
$ gcc -Wall -o test test.c && ./test
open: Too many open files
getaddrinfo: Name or service not known
```

**Comment by talex5** [comment]
Upstream glibc issue: https://sourceware.org/bugzilla/show_bug.cgi?id=29717 ("getaddrinfo returns EAI_NONAME with errno set to EMFILE")
